clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -621,56 +824,177 @@ public void assign(Collection<TopicPartition> partitions) {
+        if (!cachedSubscriptionHasAllFetchPositions && pollTimeout > retryBackoffMs) {
+            pollTimeout = retryBackoffMs;
+        }
+
+        log.trace("Polling for fetches with timeout {}", pollTimeout);
+
+        Timer pollTimer = time.timer(pollTimeout);
+
+        // Attempt to fetch any data. It's OK if we don't have any waiting data here; it's a 'best case' effort. The
+        // data may not be immediately available, but the calling method (poll) will correctly
+        // handle the overall timeout.
+        try {
+            while (pollTimer.notExpired()) {
+                CompletedFetch completedFetch = fetchResults.poll(pollTimer.remainingMs(), TimeUnit.MILLISECONDS);
+
+                if (completedFetch != null)
+                    fetchBuffer.add(completedFetch);
+
+                pollTimer.update();
+            }
+        } catch (InterruptedException e) {
+            log.trace("Timeout during fetch", e);
+        } finally {
+            timer.update(pollTimer.currentTimeMs());
+        }
+
+        return fetchCollector.collectFetch(fetchBuffer);
+    }
+
+    /**
+     * Set the fetch position to the committed position (if there is one)
+     * or reset it using the offset reset policy the user has configured.
+     *
+     * @throws org.apache.kafka.common.errors.AuthenticationException if authentication fails. See the exception for more details
+     * @throws NoOffsetForPartitionException If no offset is stored for a given partition and no offset reset policy is
+     *             defined
+     * @return true iff the operation completed without timing out
+     */
+    private boolean updateFetchPositions(final Timer timer) {
+        // If any partitions have been truncated due to a leader change, we need to validate the offsets
+        applicationEventHandler.add(new ValidatePositionsApplicationEvent());
+
+        cachedSubscriptionHasAllFetchPositions = subscriptions.hasAllFetchPositions();
+        if (cachedSubscriptionHasAllFetchPositions) return true;
+
+        // If there are any partitions which do not have a valid position and are not
+        // awaiting reset, then we need to fetch committed offsets. We will only do a
+        // coordinator lookup if there are partitions which have missing positions, so
+        // a consumer with manually assigned partitions can avoid a coordinator dependence
+        // by always ensuring that assigned partitions have an initial position.
+        if (isCommittedOffsetsManagementEnabled() && !refreshCommittedOffsetsIfNeeded(timer))
+            return false;
+
+        // If there are partitions still needing a position and a reset policy is defined,
+        // request reset using the default policy. If no reset strategy is defined and there
+        // are partitions with a missing position, then we will raise a NoOffsetForPartitionException exception.
+        subscriptions.resetInitializingPositions();
```

Comment:
There is a subtle difference between transitioning to reset from initializing and transitioning to reset from `OffsetOutOfRangeException` during fetch. In the latter, the application thread will call `FetchCollector.handleInitializeErrors()`. If there is no default offset reset policy, an `OffsetOutOfRangeException` will be thrown to the application thread during `poll`, which is what we want.

However, for the former, if there is no default offset reset policy, we simply ignore that partition through `OffsetFetcherUtils.getOffsetResetTimestamp`. It seems in that case, the partition will be forever in the reset state and the application thread won't get the `OffsetOutOfRangeException`.

Reply:
@lianetm Can you chime in on this question? Thanks!

Reply:
The issue is with how `OffsetsRequestManager.resetPositionsIfNeeded()` handles the call to `OffsetFetcherUtils.getOffsetResetTImestamp()`. Since `getOffsetResetTImestamp()` may throw an exception, `OffsetsRequestManager` should catch it and forward it to the application thread. But instead, `OffsetsRequestManager` allows the error to bubble up to the `ApplicationEventProcessor`, which will then bubble it up to `ConsumerNetworkThread`, which handles it by logging the error and exiting its `run()` method.

In fact, this appears to be the intended behavior since `OffsetsRequestManagerTest.testResetPositionsThrowsPreviousException()` explicitly tests that `OffsetsRequestManager.resetPositionsIfNeeded()` will throw the error directly at the caller.

What needs to change is for `OffsetsRequestManager.resetPositionsIfNeeded()` to catch the exception and enqueue an `ErrorBackgroundEvent` on the background event queue. In this way, the error will be processed  on the application thread as part of `poll()`.

Reply:
I don't see any existing unit tests which track the specific case that @junrao mentioned. I'm currently looking at the integration tests to see if those cover this case.

Reply:
Unfortunately, there are no unit tests or integration tests which catch the offset reset case. I'll file a Jira to have one concocted.

Reply:
I've made the following changes:

1. Now `OffsetsRequestManager` correctly enqueues errors from `getOffsetResetTimestamp()` onto the background event queue for use by the application thread
2. Renamed `OffsetFetcherTest`'s `testRestOffsetsAuthorizationFailure()` to  `testResetOffsetsAuthorizationFailure()`
3. Renamed `OffsetsRequestManagerTest`'s `testResetPositionsThrowsPreviousException()` to `testResetOffsetsAuthorizationFailure()` and updated its logic to ensure the expected error is present in the queue for the application thread to check in `poll()`

Reply:
Filed KAFKA-15652 to ensure that we have tests that catch this case.